File: traversals/src/test/scala/io/shiftleft/overflowdb/traversals/testdomains/gratefuldead/SongDb.scala
package io.shiftleft.overflowdb.traversals.testdomains.gratefuldead

import io.shiftleft.overflowdb.{NodeLayoutInformation, NodeRef, OdbNode}
import org.apache.tinkerpop.gremlin.structure.VertexProperty
import org.apache.tinkerpop.gremlin.util.iterator.IteratorUtils
import io.shiftleft.overflowdb.OdbNodeProperty

class SongDb(ref: NodeRef[SongDb]) extends OdbNode(ref) {
  private var _name: String = null
  private var _songType: String = null
  private var _performances: Integer = null

  def name: String = _name
  def songType: String = _songType
  def performances: Integer = _performances

  override protected def layoutInformation = Song.layoutInformation

  override def valueMap = {
    val properties = new java.util.HashMap[String, Any]
    if (_name != null) properties.put(Song.Properties.Name, _name)
    if (_songType != null) properties.put(Song.Properties.SongType, _songType)
    if (_performances != null) properties.put(Song.Properties.Performances, _performances)
    properties
  }
  
  override protected def specificProperties[V](key: String) = 
    key match {
      case Song.Properties.Name if _name != null => 
        IteratorUtils.of(new OdbNodeProperty(this, key, _name.asInstanceOf[V]))
      case Song.Properties.SongType if _songType != null => 
        IteratorUtils.of(new OdbNodeProperty(this, key, _songType.asInstanceOf[V]))
      case Song.Properties.Performances if _performances != null => 
        IteratorUtils.of(new OdbNodeProperty(this, key, _performances.asInstanceOf[V]))
      case _ => 
        java.util.Collections.emptyIterator
    }

  override protected def updateSpecificProperty[V](cardinality: VertexProperty.Cardinality, key: String, value: V) = 
    key match {
      case Song.Properties.Name => 
        _name = value.asInstanceOf[String]
        property(Song.Properties.Name)
      case Song.Properties.SongType => 
        _songType = value.asInstanceOf[String]
        property(Song.Properties.SongType)
      case Song.Properties.Performances => 
        _performances = value.asInstanceOf[Integer]
        property(Song.Properties.Performances)
      case _ =>       
        throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }

  override protected def removeSpecificProperty(key: String) = 
    key match {
      case Song.Properties.Name => _name = null
      case Song.Properties.SongType => _songType = null
      case Song.Properties.Performances => _performances = null
      case _ =>       
        throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
}

File: tinkerpop3/src/test/java/io/shiftleft/overflowdb/testdomains/gratefuldead/SongDb.java
package io.shiftleft.overflowdb.testdomains.gratefuldead;

import io.shiftleft.overflowdb.NodeLayoutInformation;
import io.shiftleft.overflowdb.NodeRef;
import io.shiftleft.overflowdb.OdbNode;
import io.shiftleft.overflowdb.OdbNodeProperty;
import org.apache.tinkerpop.gremlin.structure.VertexProperty;
import org.apache.tinkerpop.gremlin.util.iterator.IteratorUtils;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Map;

public class SongDb extends OdbNode {
  protected SongDb(NodeRef ref) {
    super(ref);
  }

  private String _name;
  private String _songType;
  private Integer _performances;

  public String name() {
    return _name;
  }

  public String songType() {
    return _songType;
  }

  public Integer performances() {
    return _performances;
  }

  @Override
  public NodeLayoutInformation layoutInformation() {
    return layoutInformation;
  }

  @Override
  protected <V> Iterator<VertexProperty<V>> specificProperties(String key) {
    final VertexProperty<V> ret;
    if (Song.NAME.equals(key) && _name != null) {
      return IteratorUtils.of(new OdbNodeProperty(this, key, _name));
    } else if (key == Song.SONG_TYPE && _songType != null) {
      return IteratorUtils.of(new OdbNodeProperty(this, key, _songType));
    } else if (key == Song.PERFORMANCES && _performances != null) {
      return IteratorUtils.of(new OdbNodeProperty(this, key, _performances));
    } else {
      return Collections.emptyIterator();
    }
  }

  @Override
  public Map<String, Object> valueMap() {
    Map<String, Object> properties = new HashMap<>();
    if (_name != null) properties.put(Song.NAME, _name);
    if (_songType != null) properties.put(Song.SONG_TYPE, _songType);
    if (_performances != null) properties.put(Song.PERFORMANCES, _performances);
    return properties;
  }

  @Override
  protected <V> VertexProperty<V> updateSpecificProperty(
      VertexProperty.Cardinality cardinality, String key, V value) {
    if (Song.NAME.equals(key)) {
      this._name = (String) value;
    } else if (Song.SONG_TYPE.equals(key)) {
      this._songType = (String) value;
    } else if (Song.PERFORMANCES.equals(key)) {
      this._performances = ((Integer) value);
    } else {
      throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
    return property(key);
  }


  @Override
  protected void removeSpecificProperty(String key) {
    if (Song.NAME.equals(key)) {
      this._name = null;
    } else if (Song.SONG_TYPE.equals(key)) {
      this._songType = null;
    } else if (Song.PERFORMANCES.equals(key)) {
      this._performances = null;
    } else {
      throw new RuntimeException("property with key=" + key + " not (yet) supported by " + this.getClass().getName());
    }
  }

  public static NodeLayoutInformation layoutInformation = new NodeLayoutInformation(
      1,
      new HashSet<>(Arrays.asList(Song.NAME, Song.SONG_TYPE, Song.PERFORMANCES)),
      Arrays.asList(SungBy.layoutInformation, WrittenBy.layoutInformation, FollowedBy.layoutInformation),
      Arrays.asList(FollowedBy.layoutInformation));
}
